feat(mono, app, app-shell, app-shell-odd): Add Sentry support for electron apps

mjhuff opened the pull request:
Closes [EXEC-1809](https://opentrons.atlassian.net/browse/EXEC-1809), [EXEC-1810](https://opentrons.atlassian.net/browse/EXEC-1810), and [EXEC-1811](https://opentrons.atlassian.net/browse/EXEC-1811)

<!--
Thanks for taking the time to open a Pull Request (PR)! Please make sure you've read the "Opening Pull Requests" section of our Contributing Guide:

https://github.com/Opentrons/opentrons/blob/edge/CONTRIBUTING.md#opening-pull-requests

GitHub provides robust markdown to format your PR. Links, diagrams, pictures, and videos along with text formatting make it possible to create a rich and informative PR. For more information on GitHub markdown, see:

https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax

To ensure your code is reviewed quickly and thoroughly, please fill out the sections below to the best of your ability!
-->

# Overview

This PR adds Sentry to our electron apps. Certain commits that are more interesting are highlighted below. This work is easiest to view commit-by-commit, although note that some of the mixpanel logic is changed in later commits. Later mergebacks are done to accomodate integrating the changes introduced in https://github.com/Opentrons/opentrons/pull/19561. See Review Requests, too.

On a high level, we use the `@sentry/electron` package, which contains `main` and `renderer` libraries. These libraries contain specific `integrations`, which are effectively plugins used to report certain events. The `renderer` library contains all the same `integrations` as the `@sentry/react` package that we'd care about using for the browser-layer. 

We have two separate projects the `app` and `odd`, and we use environmental tags of `production` and `development`, which make use of the existing `NODE_ENV` variable, which seems sufficient enough for our use-case. Version information (ex, `8.6.0-alpha.2`) is also included.

[8520c82](https://github.com/Opentrons/opentrons/pull/19569/commits/8520c82b5ff4d5bca822fb217a64cb19e974de4e) - Adding Sentry to our shell layers triggers an electron rebuild, which in turn led to failures due to some lower-level lib compatibility issues. Turns out we had some old dev webpack packages left around, so these are just cruft that need to be removed now.

[bee2768](https://github.com/Opentrons/opentrons/pull/19569/commits/bee2768dbe5bf297fb57abb566ffec6d0bdf9ba9) - Vite config needs some notion of whether it is a desktop or odd build target, because we keep separate sentry projects for the desktop/odd yet use the same `app` for both projects.

Note that this PR has an oe-core complimentary PR. [See that PR for details.](https://github.com/Opentrons/oe-core/pull/217)

<!--
Describe your PR at a high level. State acceptance criteria and how this PR fits into other work. Link issues, PRs, and other relevant resources.
-->

## Test Plan and Hands on Testing
- [Here is an app build](https://opentrons.slack.com/archives/C81CR4VCZ/p1758215182339179) and [here is an odd build](https://opentrons.slack.com/archives/C81CR4VCZ/p1758224496858929) based on the head of this branch. The ODD build incorporates the complementary oe-core PR, too. Running these builds with analytics enabled will report valid errors to Sentry, which are visible on the Sentry dashboard.
<!--
Describe your testing of the PR. Emphasize testing not reflected in the code. Attach protocols, logs, screenshots and any other assets that support your testing.
-->

## Changelog
- Added Sentry to deskop and odd builds.
<!--
List changes introduced by this PR considering future developers and the end user. Give careful thought and clear documentation to breaking changes.
-->

## Review requests
- I'm selecting integrations based on a combination of what I think is useful and what PD uses, but feel free to [peruse the docs](https://docs.sentry.io/platforms/javascript/guides/electron/configuration/integrations/) and suggest different options.
- We do filter out a few common errors, see the `beforeSend` callback in the app's `sentry.ts`. If there are any modifications that should be made to this list (or if we should filter any out from the shell layers, although they don't seem to be problematic), let me know.
- There is currently no gating when it comes to injecting the sentry tokens: any user who explicitly builds the app/an ot-3 system update image on CI and then runs it will cause sentry reporting (ie, custom builds). This is true of mixpanel events, too, so it's consistent. Just something to be aware of. 
<!--
- What do you need from reviewers to feel confident this PR is ready to merge?
- Ask questions.
-->

## Risk assessment
- lowish - these changes are relatively orthogonal to everything else, so I think the only thing that has a real chance of breaking here is Sentry itself
<!--
- Indicate the level of attention this PR needs.
- Provide context to guide reviewers.
- Discuss trade-offs, coupling, and side effects.
- Look for the possibility, even if you think it's small, that your change may affect some other part of the system.
  - For instance, changing return tip behavior may also change the behavior of labware calibration.
- How do your unit tests and on hands on testing mitigate this PR's risks and the risk of future regressions?
- Especially in high risk PRs, explain how you know your testing is enough.
-->

[EXEC-1809]: https://opentrons.atlassian.net/browse/EXEC-1809?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[EXEC-1810]: https://opentrons.atlassian.net/browse/EXEC-1810?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[EXEC-1811]: https://opentrons.atlassian.net/browse/EXEC-1811?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

mjhuff committed:
feat(app-shell): support sentry in app-shell

mjhuff committed:
refactor(monorepo): remove webpack dev dependencies

mjhuff committed:
feat(app-shell-odd): support sentry in app-shell-odd

mjhuff committed:
feat(app): support sentry in app

mjhuff committed:
refactor(monorepo): update yarn.lock

mjhuff committed:
feat(monorepo): inject necessary envs at build time

sfoster1 commented on app-shell/src/sentry.ts:
```diff
@@ -0,0 +1,63 @@
+    return
+  }
+
+  if (MIXPANEL_ID == null) {
```
is this really the way to check whether analytics is disabled? doesn't it have to come from config?

sfoster1 commented on app-shell/src/sentry.ts:
```diff
@@ -0,0 +1,63 @@
+  }
+
+  try {
+    init({
```
should add `net::ERR_INTERNET_DISCONNECTED` somewhere around here to be ignored

mjhuff committed:
feat(app): add error boundary reporting to sentry

mjhuff committed:
err_internet_disconnected

mjhuff committed:
get mixpanel opt in from config

mjhuff committed:
update the project names

mjhuff committed:
Merge branch 'edge' into add-sentry-electron-apps

mjhuff committed:
cleanup

mjhuff committed:
Merge branch 'edge' into add-sentry-electron-apps

mjhuff commented on .storybook/vite.config.mts:
```diff
@@ -73,6 +74,7 @@ export default defineConfig({
     _OT_PD_SENTRY_DEV_DSN_: JSON.stringify(process.env.OT_PD_SENTRY_DEV_DSN),
     _OT_PD_SENTRY_DSN_: JSON.stringify(process.env.OT_PD_SENTRY_DSN),
     _OT_PD_VERSION_: JSON.stringify(process.env.OT_PD_VERSION),
+    'process.env.NODE_DEBUG': JSON.stringify(process.env.NODE_DEBUG),
```
For some currently unexplainable reason, we need to inject this value for the app to render properly, even though this PR did not introduce granular env injection. A lib uses this value internally.

Even more confusingly, pd and labware library e2e testing _on CI only_ require that their respective `vite.config.mts` files inject this value to pass...

sfoster1 committed:
Merge branch 'edge' into add-sentry-electron-apps

sfoster1 committed:
chore(app(-shell(-odd)?)?): sentry 7.5

Maybe this will make the ODD actually work

sfoster1 committed:
Merge branch 'edge' into add-sentry-electron-apps

ddcc4 commented on app/src/App/sentry.ts:
```diff
@@ -0,0 +1,93 @@
+      integrations: [
+        ...getDefaultIntegrations(sentryOptions),
+        functionToStringIntegration(),
+        captureConsoleIntegration({ levels: ['error'] }),
```
Do you use `assert` at all in the app?

sfoster1 replied:
a couple places, though they're pretty core - is there an integration we're missing for them?

ddcc4 replied:
Oh, I meant: If you use `console.assert()`, you might as well add `'assert'` to `levels` here so that they get reported to Sentry.

ddcc4 commented on app/src/App/sentry.ts:
```diff
@@ -0,0 +1,93 @@
+    })
+
+    isSentryInitialized = true
+    console.log('Sentry initialized successfully.')
```
Is this `console.log()` copied from existing code? I personally find it a bit noisy to see this at the top of the console, and we'd probably be fine if we took it out.

sfoster1 committed:
catch some annoying devtools errors

sfoster1 committed:
disable sentry gpu integration

It doesn't work on this platform, which you only get to find out with
debug: true which i definitely knew about and hadn't forgotten.

sfoster1 committed:
remove noisy log

sfoster1 committed:
update filters for server-not-up

sfoster1 committed:
sentry hooks - allow captureMessage
